### PR TITLE
add opensearch script for a developing purpose

### PR DESCRIPTION
**Pull request by carlory**
Signed-off-by: carlory <baofa.fan@daocloud.io>

**What type of PR is this?**

/kind feature

**Which issue(s) this PR fixes**:
Part of #1794

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
add opensearch script for a developing purpose
```

### CODE REVIEW

File: artifacts/opensearch/karmada-opensearch-dashboards.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  namespace: karmada-system
+spec:
+  ports:
+  - name: http
+    port: 5601
+    protocol: TCP
+    targetPort: 5601
+  selector:
+    app: karmada-opensearch-dashboards
+  type: ClusterIP
```
Comment (Poor12) [inline artifacts/opensearch/karmada-opensearch-dashboards.yaml:46]: lint: add a new line

File: artifacts/opensearch/karmada-opensearch.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+spec:
+  ports:
+  - name: client
+    port: 9200
+    targetPort: 9200
+  - name: analyzer
+    port: 9600
+    targetPort: 9600
+  selector:
+     app: karmada-opensearch
```
Comment (Poor12) [inline artifacts/opensearch/karmada-opensearch.yaml:64]: lint: add a new line

**Review by Poor12 — COMMENTED** [review]
Could you show some test results?

**Comment by RainbowMango** [comment]
Thank you very much @carlory, please give an example of how to use this script, the reviewer(such as me) would like to try it.

cc @liys87x here to review.

**Comment by carlory** [comment]
Examples (script):

```
(⎈ |karmada:demo)➜  karmada git:(add-opensearch-script) hack/deploy-karmada-opensearch.sh ~/.kube/config ik8s01
Switched to context "ik8s01".
using kubeconfig: /var/folders/n6/33y75ytn36g38k2wnkb6lp8r0000gn/T/tmp.D9s9xoof/kubeconfig
deployment.apps/karmada-opensearch created
service/karmada-opensearch created
deployment.apps/karmada-opensearch-dashboards created
service/karmada-opensearch-dashboards created
wait the karmada-opensearch ready...
pod/karmada-opensearch-7c7455d8dc-bsfzb condition met
wait the karmada-opensearch-dashboards ready...
pod/karmada-opensearch-dashboards-566465cf5b-jpnhf condition met
Opensearch is deployed successfully.
You can access the opensearch at http://karmada-opensearch.karmada-system.svc:9200

(⎈ |karmada:demo)➜  karmada git:(add-opensearch-script) kubectl --context ik8s01 get pods
NAME                                               READY   STATUS    RESTARTS        AGE
karmada-opensearch-7c7455d8dc-bsfzb                1/1     Running   0               75s
karmada-opensearch-dashboards-566465cf5b-jpnhf     1/1     Running   0               75s
...
```

ResourceRegistry  Examples:

```
 apiVersion: search.karmada.io/v1alpha1
 kind: ResourceRegistry
 metadata:
   name: foo
 spec:
   backendStore:
     openSearch:
       addresses:
       - http://karmada-opensearch.karmada-system.svc:9200
       secretRef:
         name: ""
         namespace: ""
   resourceSelectors:
   - apiVersion: apps/v1
     kind: Deployment
   targetCluster:
     clusterNames:
     - k8s
     - ocp
```

**Comment by carlory** [comment]
visit opensearch dashboard (http://127.0.0.1:5601):

```shell
(⎈ |ik8s01:karmada-system)➜  karmada git:(add-opensearch-script) kubectl port-forward svc/karmada-opensearch-dashboards 5601:5601
Forwarding from 127.0.0.1:5601 -> 5601
Forwarding from [::1]:5601 -> 5601
```
image

**Comment by carlory** [comment]
image

image

https://github.com/karmada-io/karmada/blob/cc74a423222ccf1b6896d7fa8c314544a24cade7/pkg/search/backendstore/opensearch.go#L30

I'm not sure if there is a compatibility issue here

/cc @RainbowMango @Poor12 @liys87x

It doesn't seem to be a problem with the script.

**Comment by liys87x** [comment]
> This is the problem with the code
https://github.com/karmada-io/karmada/blob/cc74a423222ccf1b6896d7fa8c314544a24cade7/pkg/search/backendstore/opensearch.go#L228

**Comment by carlory** [comment]
@liys87x 👍

thanks for you guide, I will fill another pr to fix it.

**Comment by liys87x** [comment]
Thank you very much @carlory, I have tried it, It improves our experience. @RainbowMango

File: artifacts/opensearch/karmada-opensearch.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        fsGroup: 1000
+        runAsGroup: 1000
+        runAsUser: 1000
+      volumes:
+      - emptyDir: {}
+        name: karmada-opensearch
+---
+apiVersion: v1
+kind: Service
+metadata:
```
Comment (liys87x) [inline artifacts/opensearch/karmada-opensearch.yaml:53]: namespace not specified

File: artifacts/opensearch/karmada-opensearch.yaml [hunk 4]
```diff
@@ -0,0 +1,66 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
```
Comment (liys87x) [inline artifacts/opensearch/karmada-opensearch.yaml:3]: namespace not specified

File: hack/deploy-karmada-opensearch.sh [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+ # switch to host cluster
+TEMP_PATH=$(mktemp -d)
+cp $HOST_CLUSTER_KUBECONFIG $TEMP_PATH/kubeconfig
+export KUBECONFIG="$TEMP_PATH/kubeconfig"
+kubectl config use-context "${HOST_CLUSTER_NAME}"
+echo "using kubeconfig: "$KUBECONFIG
+
+# deploy karmada opensearch
+kubectl apply -f "${REPO_ROOT}/artifacts/opensearch/karmada-opensearch.yaml"
```
Comment (liys87x) [inline hack/deploy-karmada-opensearch.sh:49]: namespace not specified

File: hack/deploy-karmada-opensearch.sh [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+kubectl apply -f "${REPO_ROOT}/artifacts/opensearch/karmada-opensearch-dashboards.yaml"
+
+# make sure all karmada-opensearch components are ready
+util::wait_pod_ready "${KARMADA_OPENSEARCH_LABEL}" "${KARMADA_SYSTEM_NAMESPACE}"
```
Comment (liys87x) [inline hack/deploy-karmada-opensearch.sh:53]: The namespace is specified here, but these resources are created in the `default` namespace

**Review by liys87x — COMMENTED** [review]
Inconsistent namespace @carlory

**Comment by carlory** [comment]
@liys87x updated

**Comment by liys87x** [comment]
We don't want to use dynamic mapping, it can lead to wrong type. Do you have any suggestion. @carlory 
https://github.com/karmada-io/karmada/blob/cc74a423222ccf1b6896d7fa8c314544a24cade7/pkg/search/backendstore/opensearch.go#L86-L91

**Comment by carlory** [comment]
I don't know much about this, I need to investigate it. 😁

**Comment by RainbowMango** [comment]
Thanks. I'll try it on my side today.

**Comment by RainbowMango** [comment]
/approve 
Thanks. carlory

**Pull request merged**